### PR TITLE
chore: add memory report into the CI

### LINKED ISSUE #6604: Integrate Heapstack into CI to track the peak memory usage of the compiler on a set of programs.
Heaptrack has proven to be a useful tool so we should try to automate its usage. One easy way to do this is to add tracking of the peak memory usage in PRs, any significant changes should be reported with a warning comment.

We shouldn't need too many programs for this: a small, medium and large one should suffice as it'll allow us to see the fixed size memory costs vs worst case.

Ideally this should emit a workflow artifact of the profiling run so we can easily just download it and open in heaptrack.

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Resolves #6604 

## Summary\*
Add a bash script which computes memory peak usage on 3 test programs, and a CI action to include this information on CI.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by guipublic** [comment]
In noir-diff, I use a workaround to not download the 'previous artefact', because for the memory report, this part stalls, and I don't know why.
As a result, I do not do a diff with the previous artefact.

**Comment by TomAFrench** [comment]
>  because for the memory report, this part stalls, and I don't know why.

This is because there's no report in the master branch and so it's just looking back through years of git history for one. We could add a bootstrap commit option to the action to not look past a certain commit.

**Comment by guipublic** [comment]
Peak memory profiling for test programs show that most of the programs are about 80M of peak memory usage, except for the ones below. 
For the report I chose:
- Keccak as a small peak, which I think is a good candidate for representing what Noir can be used for
- Workspace for a medium peak, which could indicate frontend issues
- regression_4709 for a higher peak. The test is simple, it is doing a many additions in a nested loop.


| Program                                | Peak Memory |
| -------------------------------------- | ----------- |
| bench_2_to_17                          | 3.1G        |
| regression_5252                        | 3G          |
| ram_blowup_regression                  | 2.8G        |
| eddsa                                  | 885M        |
| poseidonsponge_x5_254                  | 396M        |
| regression_4709                        | 387M        |
| hashmap                                | 280M        |
| sha256_var_padding_regression          | 237M        |
| sha256_regression                      | 190M        |
| regression_4449                        | 185M        |
| sha2_byte                              | 150M        |
| sha256_var_size_regression             | 130M        |
| workspace                              | 128M        |
| poseidon_bn254_hash_width_3            | 128M        |
| conditional_1                          | 115M        |
| uhashmap                               | 114M        |
| no_predicates_numeric_generic_poseidon | 104M        |
| fold_numeric_generic_poseidon          | 102M        |
| sha256                                 | 93M         |
| keccak256                              | 90M         |
| debug_logs                             | 88M         |
| array_dynamic_blackbox_input           | 87M         |
| nested_array_dynamic                   | 86M         |
| sha256_var_witness_const_regression    | 84M         |

File: .github/workflows/memory_report.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          save-if: ${{ github.event_name != 'merge_group' }}
+
+      - name: Build Nargo
+        run: cargo build --package nargo_cli --release
+
+      - name: Package artifacts
+        run: |
+          mkdir dist
+          cp ./target/release/nargo ./dist/nargo
+          7z a -ttar -so -an ./dist/* | 7z a -si ./nargo-x86_64-unknown-linux-gnu.tar.gz
```
Comment (TomAFrench) [inline .github/workflows/memory_report.yml:36]:
```suggestion
```

File: .github/workflows/memory_report.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      - name: Generate Memory report
+        working-directory: ./test_programs
+        run: |
+          chmod +x memory_report.sh
+          ./memory_report.sh
+          mv memory_report.json ../memory_report.json
+
+      - name: Parse memory report
+        id: memory_report
+        uses: noir-lang/noir-bench-report@91003b0bfe46c4e42c1714f9450917d0103c57ca
```
Comment (TomAFrench) [inline .github/workflows/memory_report.yml:77]:
```suggestion
        uses: noir-lang/noir-bench-report@d61bc78ece3c8df1e72914b3d5136bad403d5bdf
```

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**